nixos/filesystems: add automatic nilfs2 detection

rapenne-s opened the pull request:
###### Description of changes

This adds a detection for nilfs2 filesystems in mountpoints, and automatically pulls in the kernel module. Without it, you need to add the `nilfs2` kernel module to have a successful stage-1, otherwise you obtain a "device not found" at boot which isn't very helpful. Successfully tested on a laptop using nilfs2 for its root filesystem, no extra configuration required to boot it :rocket: 

fixes #65548

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ckiee commented:
Haven't been doing much nixpkgs work lately, sorry. Maybe post on [the discourse thread](https://discourse.nixos.org/t/prs-ready-for-review/3032)?

rapenne-s commented:
> Haven't been doing much nixpkgs work lately, sorry. Maybe post on [the discourse thread](https://discourse.nixos.org/t/prs-ready-for-review/3032)?

I'll try that, thank you :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1223

rapenne-s commented:
I fixed another issue, nilfs2 filesystem require a program named `nilfs_cleanerd` to run to handle the garbage collection, but it was failing. It requires a `/etc/nilfs_cleanerd.conf` configuration file. Just create an empty one when there is a nilfs filesystem, and I added a simple nixos module to make it configurable.

rapenne-s commented:
@S-NA maybe this interest you? :) I saw you made changes to nilfs-utils "recently"

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/solene-linux-nilfs-file-system-automatic-continuous-snapshots/22246/1

aanderse commented on nixos/modules/profiles/base.nix:
```diff
@@ -51,7 +51,7 @@
 
   # Include support for various filesystems.
-  boot.supportedFilesystems = [ "btrfs" "reiserfs" "vfat" "f2fs" "xfs" "zfs" "ntfs" "cifs" ];
+  boot.supportedFilesystems = [ "btrfs" "reiserfs" "vfat" "f2fs" "nilfs2" "xfs" "zfs" "ntfs" "cifs" ];
```
I'm not sure we want to add this by default. I liked your blog post on this software but I think we should probably only add items here that are a bit more popular.

My memory on the filesystem code is a bit hazy... but would it create an infinite recursion here if we conditionally added this value depending on whether the user has added this file system to their system via `config.fileSystems`?

rapenne-s replied:
I saw reiserfs so I thought it would be fine. I can try to add it conditionally and see if it works.

rapenne-s replied:
I removed it from the list, it's not useful to boot on a nilfs root filesystem. It seems to be used to provide tools in the installer ISO and I don't think we need to have nilfs-utils by default.

aanderse commented on nixos/modules/tasks/filesystems/nilfs2.nix:
```diff
@@ -0,0 +1,57 @@
+  options = {
+    boot.nilfs2 = {
+      cleanerd_conf = mkOption {
+        type = types.lines;
```
Any interest in using an attribute set here instead?

rapenne-s replied:
This would be better indeed

rapenne-s replied:
it's using attributes now, way better :+1:

Commit:
nixos/filesystems: add automatic nilfs2 detection

aanderse commented on nixos/modules/tasks/filesystems/nilfs2.nix:
```diff
@@ -0,0 +1,63 @@
+  options = {
+    boot.nilfs2 = {
+      cleanerd_conf = mkOption {
+        type = types.attrsOf types.lines;
```
Would `(pkgs.formats.keyValue { }).format` work here?

aanderse commented on nixos/modules/tasks/filesystems/nilfs2.nix:
```diff
@@ -0,0 +1,63 @@
+  inInitrd = any (fs: fs == "nilfs2") config.boot.initrd.supportedFilesystems;
+
+  # this file is mandatory for nilfs_cleanerd - can be empty
+  cleaner_conf = pkgs.writeText "nilfs_cleanerd.conf" ''
```
Would `(pkgs.formats.keyValue { }).generate "cleaner.conf" config.boot.nilfs2.cleanerd_conf` work here?
